src/native/corehost/hostmisc/trace.cpp

```diff
@@ -13,9 +13,21 @@
 //  COREHOST_TRACE=1 COREHOST_TRACE_VERBOSITY=1          implies g_trace_verbosity = 1.  // Trace "enabled".  error() messages will be produced
 static int g_trace_verbosity = 0;
 static FILE * g_trace_file = stderr;
-static pal::mutex_t g_trace_mutex;
+static pal::mutex_t* g_trace_mutex;
 thread_local static trace::error_writer_fn g_error_writer = nullptr;
 
+namespace
+{
+    pal::mutex_t& get_mutex()
+    {
+        static pal::mutex_t* mutex = new pal::mutex_t();
+        if (g_trace_mutex == nullptr)
+            g_trace_mutex = mutex;
```

Comment:
For normal hosting operations this is probably OK - the only scenario where this might be problematic is a tool using hosting APIs to probe for things (would never actually start the runtime - basically just using either nethost, or hostfxr which are both cleanly unloadable). If such tool does this repeatedly and it causes errors, each such call would leak a mutex.

Is there really no way to clean up?

One possible idea (but I'm no C++ expert): custom class with a destructor which wraps the mutex. In the destructor it tries to enter the mutex (with a timeout) - if that succeeds, it knows it's the only one there and can delete it and cleanup the wrapper. If it fails - let it leak (should basically never happen). Would probably need some double check when entering the mutex (enter the mutex, check that the wrapper hasn't been destroyed, and only then continue).

I'm just a bit nervous intentionally leaking stuff - especially in the nethost and hostfxr which are meant to be unloadable.

Reply:
Another option would be to use a simple spin-lock and remove the `mutex` logic entirely. Unsure how long these ops are supposed to take, but it might be an okay compromise.

One could also simply create a "private" one and leave the public a pointer as done above. This way cleanup would still happen.

```cpp
static pal::mutex_t g_trace_mutex_private;
static pal::mutex_t* g_trace_mutex;
...
    pal::mutex_t& get_mutex()
    {
        if (g_trace_mutex == nullptr)
            g_trace_mutex = &g_trace_mutex_private;

        return *g_trace_mutex;
    }
```

Reply:
Oops. The second option isn't tenable I see. Avoiding destruction entirely is the point. Hmmm spin-lock seems appropriate to me then.

Reply:
Yeah, I (obviously) had a hard time figuring out how to clean up for the library case. I think any sort of destructor would land us in the same place - the object gets destructed and then some other thread calls into it, resulting in undefined behaviour / badness.

I guess rolling our own spin lock might be it... These are tracing operations, so how long it takes kind of depends on I/O as we write to a file/stdout/stderr.

Reply:
The spin lock is likely something that can be inefficient if this is only for tracing since it is by definition for investigating a bug or issue so speed shouldn't be the priority, I think?

Reply:
Yeah. I don't think speed is the priority. The non-error, non-investigation scenario for tracing would be commands for printing information - for example, `--info`, `--list-runtimes`, `--list-sdks`.

Reply:
Here is my new suggestion that does not leak any memory but still does not crash. I am not a dotnet runtime developer, so please adjust the code-style as required:

Keep '''static pal::mutex_t g_trace_mutex;''' as is.

At the top of the original trace.cpp, add this:
```

static bool g_trace_mutex_shutdown = false;


class MutexAccessGuardForCRTShutdownSituations {
public:
    MutexAccessGuardForCRTShutdownSituations() {
        g_trace_mutex_shutdown = false;
    }
    ~MutexAccessGuardForCRTShutdownSituations() {
        g_trace_mutex_shutdown = true;
    }
} g_MutexAccessGuardForCRTShutdownSituations;


```

And then add this check for g_trace_mutex_shutdown before using the mutex i.e. 
```
void trace::flush()
{
    if (g_trace_mutex_shutdown) {
        return;
    }
    
    std::lock_guard<pal::mutex_t> lock(g_trace_mutex);
```

This strange class MutexAccessGuardForCRTShutdownSituations will be destroyed when the CRT shuts down and then it will not access the mutex any more and simply do nothing, as it would be wrong to do anything anyways in that situation

Reply:
How do you guard against race conditions - what if a thread gets past the boolean check and then another thread initiates the shutdown. By the time the first thread gets to the lock the mutex might be destroyed.

The spin-lock solution is not perfect, but it's highly unlikely to cause problems:
* Tracing is almost always off - so it will never actually run
* Vast majority of the hosting code runs on a single thread, so the contention on the lock should be basically non-existent

Reply:
I'm going to stick with the spin lock. As @vitek-karas said, while it is not ideal, this is a path where we don't expect it to cause issues.